src/scale.jl

```diff
@@ -251,45 +253,51 @@ function apply_scale_typed!(ds, field, scale::ContinuousScale)
 end
 
 # Reorder the levels of a pooled data array
-function reorder_levels(da::PooledDataArray, order::AbstractVector)
-    level_values = levels(da)
+function reorder_levels(da::IndirectArray, order::AbstractVector)
+    level_values = da.values
     length(order) != length(level_values) &&
             error("Discrete scale order is not of the same length as the data's levels.")
     permute!(level_values, order)
-    return PooledDataArray(da, level_values)
+    return IndirectArray(da, level_values)
 end
 
-function discretize_make_pda(values::Vector, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(values)
-    else
-        return PooledDataArray(convert(Vector{eltype(levels)}, values), levels)
-    end
-end
+discretize_make_pda(values::Vector)         = IndirectArray(values)
+discretize_make_pda(values::Vector, ::Void) = IndirectArray(values)
+discretize_make_pda(values::Vector, levels) =
+    IndirectArray(map(t -> findfirst(levels, t), values), levels)
 
-function discretize_make_pda(values::DataArray, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(values)
-    else
-        return PooledDataArray(convert(DataArray{eltype(levels)}, values), levels)
-    end
-end
+discretize_make_pda(values::DataArray)         = IndirectArray(values)
+discretize_make_pda(values::DataArray, ::Void) = IndirectArray(values)
+discretize_make_pda(values::DataArray, levels) =
+    IndirectArray(convert(DataArray{eltype(levels)}, values), levels)
 
-function discretize_make_pda(values::Range, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(collect(values))
-    else
-        return PooledDataArray(collect(values), levels)
-    end
-end
+discretize_make_pda(values::Range)         = IndirectArray(collect(values))
+discretize_make_pda(values::Range, ::Void) = IndirectArray(collect(values))
+discretize_make_pda(values::Range, levels) = IndirectArray(collect(values), levels)
 
-function discretize_make_pda(values::PooledDataArray, levels=nothing)
-    if levels == nothing
-        return values
-    else
-        return PooledDataArray(values, convert(Vector{eltype(values)}, levels))
-    end
+discretize_make_pda(values::IndirectArray)         = values
+discretize_make_pda(values::IndirectArray, ::Void) = values
+discretize_make_pda(values::IndirectArray, levels) =
+    IndirectArray(values, convert(Vector{eltype(values)}, levels))
+
+discretize_make_pda(values::CategoricalArray)         = discretize_make_pda(values, unique(values))
```

Comment:
The ordering returned by `levels(values)` should be preserved here, that's the whole point of categorical arrays. `intersect(vcat(levels(values), missing), unique(values))` should work and will sort missing values at the end.

Reply:
This code path is not preserving order. There is a separate branch in `discretize` taken when the ordering should be preserved. We might want to change this later but I don't think such a reorganization belongs here.

Reply:
The existing `PooledDataArray` method did preserve order since it passed the `levels` argument to the PDA constructor. So I think you should keep the current `level=nothing` pattern, using the code I posted above to fill `levels` when it's `nothing`.